### PR TITLE
Local Variable caching work.

### PR DESCRIPTION
**Pull request by bretthoerner**
The work in this PR came out of lots of testing of the agent with an example Spring app.

I'd say you can trust the C++ code, I've done a bunch of testing and this is still only leading toward a beta. Mostly curious on any "wtf why" type stuff since none of the code here should adversely affect people who *aren't* using the agent.

* Add `FrameCache.shouldCacheThrowable` method, called by native code to decide whether to do further processing/caching. If not, we can exit early and avoid collecting/caching every frame's locals.
* I noticed an uncaught Spring exception would be re-thrown and caught multiple (like 5+?) times before being handed to the first exception handler I control (Sentry), with slimming stack frames *on the same exception object*. I haven't dug into what they're doing there, but I needed to adjust the `FrameCache` to not process/cache an *existing* exception that is thrown with a *smaller* stacktrace.
* "in app" frame configuration is required, if an exception is thrown that doesn't have a single app frame in it, we don't process/cache it. I'm unsure about this one long term...
* Add `Method` to each `Frame` in the cache so I can do a best-effort to match up frames in cases where a smaller/larger stacktrace is actually captured for Sentry.

In short, we now cache the "largest" stacktrace for a given exception instance, and we do a best effort (using method names) to match up the cached frames with the frames that are actually sent to Sentry.

### CODE REVIEW

**Comment by bretthoerner** [comment]
Should have said: once this PR is merged (as it fixes real bugs in real programs) and documentation (working) I think I'm ready to make an "actual" beta to point some randos at.

File: sentry/src/main/java/io/sentry/DefaultSentryClientFactory.java [hunk 1]
```diff
@@ -512,6 +513,10 @@ protected ContextManager getContextManager(Dsn dsn) {
             }
         }
 
+        for (String inAppPackage : inAppPackages) {
+            FrameCache.addAppPackage(inAppPackage);
+        }
```
Comment (tkaemming) [inline sentry/src/main/java/io/sentry/DefaultSentryClientFactory.java]: This implicit behavior updating global state seems a little weird but I haven't put a lot of thought into it — I'd expect it doesn't really cause any issues in practice but it does smell a little weird

File: sentry/src/main/java/io/sentry/event/interfaces/SentryStackTraceElement.java [hunk 2]
```diff
@@ -116,9 +116,24 @@ public String getPlatform() {
     public static SentryStackTraceElement[] fromStackTraceElements(StackTraceElement[] stackTraceElements,
                                                                    Frame[] cachedFrames) {
         SentryStackTraceElement[] sentryStackTraceElements = new SentryStackTraceElement[stackTraceElements.length];
-        for (int i = 0; i < stackTraceElements.length; i++) {
-            sentryStackTraceElements[i] = fromStackTraceElement(stackTraceElements[i],
-                cachedFrames != null ? cachedFrames[i].getLocals() : null);
+        for (int i = 0, j = 0; i < stackTraceElements.length; i++, j++) {
```
Comment (tkaemming) [inline sentry/src/main/java/io/sentry/event/interfaces/SentryStackTraceElement.java]: Comment here explaining the rationale behind this new implementation would be helpful for future readers — it makes sense in context but I think without the knowledge of this change it'd be pretty arcane

File: sentry/src/main/java/io/sentry/DefaultSentryClientFactory.java (hunk 1 above)
Comment (bretthoerner) [inline sentry/src/main/java/io/sentry/DefaultSentryClientFactory.java]: You're right, it shouldn't be in the getter.

File: sentry/src/main/java/io/sentry/event/interfaces/SentryStackTraceElement.java (hunk 2 above)
Comment (bretthoerner) [inline sentry/src/main/java/io/sentry/event/interfaces/SentryStackTraceElement.java]: Good catch, I meant to do that. 😊

Comment (tkaemming) [inline sentry/src/main/java/io/sentry/event/interfaces/SentryStackTraceElement.java]: 👍
